File: nexus_api/security_api.py
"""
NexusAgent 安全测试服务
封装 Xray + Rad 漏洞扫描引擎
"""
import asyncio
import json
import os
import subprocess
import time
import uuid
from pathlib import Path
from typing import Any, Dict, List, Optional

from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel

# ─────────────────────────────────────────────────────────────────────────────
# 模型定义
# ─────────────────────────────────────────────────────────────────────────────

class ScanRequest(BaseModel):
    target_url: str
    scan_type: str = "quick"  # quick, full, xss, sqli, cmd


class VulnInfo(BaseModel):
    id: str
    name: str
    severity: str  # critical, high, medium, low, info
    url: str
    vuln_type: str
    detail: str
    payload: Optional[str] = None
    timestamp: str


class ScanStatus(BaseModel):
    task_id: str
    status: str  # pending, running, completed, failed
    progress_percent: int
    vulns_found: int
    message: str


class ScanSummary(BaseModel):
    task_id: str
    target_url: str
    scan_type: str
    status: str
    total: int
    severity_counts: Dict[str, int]
    duration_ms: int


# ─────────────────────────────────────────────────────────────────────────────
# 扫描引擎配置
# ─────────────────────────────────────────────────────────────────────────────

SECURITY_DIR = Path(__file__).parent.parent / "security"
XRAY_PATH = SECURITY_DIR / "Xray_Rad_complete" / "xray_windows_amd64.exe"
RAD_PATH = SECURITY_DIR / "Xray_Rad_complete" / "rad_windows_amd64.exe"


# ─────────────────────────────────────────────────────────────────────────────
# 扫描任务存储
# ─────────────────────────────────────────────────────────────────────────────

_scan_tasks: Dict[str, Dict[str, Any]] = {}


# ─────────────────────────────────────────────────────────────────────────────
# FastAPI 应用
# ─────────────────────────────────────────────────────────────────────────────

def create_app() -> FastAPI:
    app = FastAPI(
        title="NexusAgent Security Scan Service",
        description="Xray + Rad 漏洞扫描服务",
        version="1.0.0",
    )

    app.add_middleware(
        CORSMiddleware,
        allow_origins=["*"],
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )

    @app.get("/health")
    async def health():
        return {"status": "ok", "service": "security"}

    @app.post("/scan")
    async def start_scan(request: ScanRequest):
        """启动安全扫描"""
        task_id = str(uuid.uuid4())[:8]

        # 初始化任务状态
        _scan_tasks[task_id] = {
            "target_url": request.target_url,
            "scan_type": request.scan_type,
            "status": "pending",
            "progress": 0,
            "vulns": [],
            "start_time": time.time(),
            "end_time": None,
        }

        # 异步执行扫描
        asyncio.create_task(_run_scan(task_id, request))

        return {"success": True, "task_id": task_id}

    @app.get("/scan/{task_id}/status")
    async def get_scan_status(task_id: str):
        """获取扫描状态"""
        if task_id not in _scan_tasks:
            return {"error": "Task not found"}

        task = _scan_tasks[task_id]
        elapsed = int((time.time() - task["start_time"]) * 1000)

        return ScanStatus(
            task_id=task_id,
            status=task["status"],
            progress_percent=task["progress"],
            vulns_found=len(task["vulns"]),
            message="扫描进行中" if task["status"] == "running" else "扫描完成",
        )

    @app.get("/scan/{task_id}/result")
    async def get_scan_result(task_id: str):
        """获取扫描结果"""
        if task_id not in _scan_tasks:
            return {"success": False, "error": "Task not found"}

        task = _scan_tasks[task_id]
        return {
            "success": True,
            "vulns": task["vulns"],
            "summary": {
                "total": len(task["vulns"]),
                "by_severity": _count_by_severity(task["vulns"]),
            },
        }

    @app.get("/scans")
    async def list_scans():
        """列出所有扫描任务"""
        scans = []
        for task_id, task in _scan_tasks.items():
            scans.append(ScanSummary(
                task_id=task_id,
                target_url=task["target_url"],
                scan_type=task["scan_type"],
                status=task["status"],
                total=len(task["vulns"]),
                severity_counts=_count_by_severity(task["vulns"]),
                duration_ms=int((task["end_time"] or time.time()) - task["start_time"]) * 1000 if task["end_time"] else 0,
            ))
        return {"scans": scans}

    @app.delete("/scan/{task_id}")
    async def delete_scan(task_id: str):
        """删除扫描记录"""
        if task_id in _scan_tasks:
            del _scan_tasks[task_id]
            return {"success": True}
        return {"success": False, "error": "Task not found"}

    @app.get("/scan/{task_id}/export")
    async def export_report(task_id: str, format: str = "json"):
        """导出扫描报告"""
        if task_id not in _scan_tasks:
            return {"success": False, "error": "Task not found"}

        task = _scan_tasks[task_id]

        if format == "json":
            report = json.dumps({
                "task_id": task_id,
                "target_url": task["target_url"],
                "scan_type": task["scan_type"],
                "vulns": task["vulns"],
                "summary": {
                    "total": len(task["vulns"]),
                    "by_severity": _count_by_severity(task["vulns"]),
                },
            }, ensure_ascii=False, indent=2)
        else:
            # HTML 格式
            report = _generate_html_report(task_id, task)

        return {"success": True, "report_content": report}

    return app


# ─────────────────────────────────────────────────────────────────────────────
# 扫描执行
# ─────────────────────────────────────────────────────────────────────────────

async def _run_scan(task_id: str, request: ScanRequest):
    """执行扫描任务"""
    task = _scan_tasks[task_id]
    task["status"] = "running"

    try:
        # 模拟扫描（实际项目中调用 Xray + Rad）
        # 这里提供简化的实现，实际部署时替换为真实扫描逻辑
        
        await asyncio.sleep(1)  # 模拟初始化
        task["progress"] = 10

        # 模拟扫描过程
        for i in range(10, 100, 10):
            await asyncio.sleep(0.5)
            task["progress"] = i

        # 模拟发现漏洞（实际项目中解析 Xray 输出）
        if "sql" in request.target_url.lower() or "sqli" in request.scan_type:
            task["vulns"].append({
                "id": f"{task_id}-1",
                "name": "SQL 注入",
                "severity": "critical",
                "url": request.target_url,
                "vuln_type": "SQL Injection",
                "detail": "检测到 SQL 注入漏洞，可能被用于窃取数据库信息",
                "payload": "' OR '1'='1",
                "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
            })

        if "xss" in request.target_url.lower() or request.scan_type == "xss":
            task["vulns"].append({
                "id": f"{task_id}-2",
                "name": "XSS 跨站脚本",
                "severity": "high",
                "url": request.target_url,
                "vuln_type": "Cross-Site Scripting",
                "detail": "检测到反射型 XSS 漏洞",
                "payload": "<script>alert(1)</script>",
                "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
            })

        task["status"] = "completed"
        task["progress"] = 100
        task["end_time"] = time.time()

    except Exception as e:
        task["status"] = "failed"
        task["progress"] = 0
        task["end_time"] = time.time()


def _count_by_severity(vulns: List[Dict]) -> Dict[str, int]:
    """统计各严重级别的漏洞数量"""
    counts = {"critical": 0, "high": 0, "medium": 0, "low": 0, "info": 0}
    for v in vulns:
        sev = v.get("severity", "info").lower()
        if sev in counts:
            counts[sev] += 1
    return counts


def _generate_html_report(task_id: str, task: Dict) -> str:
    """生成 HTML 报告"""
    vulns_html = ""
    for v in task["vulns"]:
        vulns_html += f"""
        <tr>
            <td>{v['name']}</td>
            <td class="severity-{v['severity']}">{v['severity'].upper()}</td>
            <td>{v['url']}</td>
            <td>{v['vuln_type']}</td>
            <td>{v['timestamp']}</td>
        </tr>
        """

    return f"""
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="utf-8">
        <title>安全扫描报告 - {task_id}</title>
        <style>
            body {{ font-family: -apple-system, BlinkMacSystemFont, sans-serif; padding: 40px; background: #1a1a2e; color: #eee; }}
            h1 {{ color: #ef4444; }}
            table {{ width: 100%; border-collapse: collapse; margin-top: 20px; }}
            th, td {{ padding: 12px; text-align: left; border-bottom: 1px solid #333; }}
            th {{ background: #16213e; }}
            .severity-critical {{ color: #ef4444; font-weight: bold; }}
            .severity-high {{ color: #f97316; font-weight: bold; }}
            .severity-medium {{ color: #eab308; }}
            .severity-low {{ color: #3b82f6; }}
            .severity-info {{ color: #6b7280; }}
        </style>
    </head>
    <body>
        <h1>🛡️ 安全扫描报告</h1>
        <p><strong>任务ID:</strong> {task_id}</p>
        <p><strong>目标URL:</strong> {task['target_url']}</p>
        <p><strong>扫描类型:</strong> {task['scan_type']}</p>
        <p><strong>发现漏洞:</strong> {len(task['vulns'])} 个</p>
        <table>
            <tr><th>漏洞名称</th><th>严重级别</th><th>URL</th><th>类型</th><th>发现时间</th></tr>
            {vulns_html}
        </table>
    </body>
    </html>
    """


# ─────────────────────────────────────────────────────────────────────────────
# 入口
# ─────────────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(create_app(), host="0.0.0.0", port=9244)

File: frontend/src/lib/security-api.ts
import axios from 'redaxios';

export interface SecurityVuln {
  id: string;
  name: string;
  target: string;
  severity: 'critical' | 'high' | 'medium' | 'low' | 'info';
  vuln_type: string;
  url: string;
  detail: string;
  payload?: string;
  timestamp: string;
}

export interface SecurityScanResult {
  success: boolean;
  task_id: string;
  elapsed_ms: number;
  target_url: string;
  scan_type: string;
  vuln_count: number;
  vulns: SecurityVuln[];
  stdout?: string;
  stderr?: string;
  error?: string;
}

export interface SecurityScanSummary {
  task_id: string;
  target_url: string;
  scan_type: string;
  severity_counts: {
    critical: number;
    high: number;
    medium: number;
    low: number;
    info: number;
  };
  total: number;
  scanned_at: string;
  duration_ms: number;
}

// ============ Security Module API Client ============

const SECURITY_BASE = 'http://127.0.0.1:9244';

async function apiRequest<T>(path: string, options?: RequestInit): Promise<T> {
  const url = path.startsWith('http') ? path : `${SECURITY_BASE}${path}`;
  const response = await fetch(url, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers,
    },
  });
  if (!response.ok) {
    throw new Error(`Security API error: ${response.status}`);
  }
  return response.json();
}

export async function startSecurityScan(params: {
  target_url: string;
  scan_type: 'quick' | 'full' | 'xss' | 'sqli' | 'cmd';
  enable_auth?: boolean;
  auth_username?: string;
  auth_password?: string;
}): Promise<{ success: boolean; task_id: string; message?: string; error?: string }> {
  return apiRequest('/scan', {
    method: 'POST',
    body: JSON.stringify(params),
  });
}

export async function getSecurityScanStatus(taskId: string): Promise<{
  success: boolean;
  status: 'pending' | 'running' | 'completed' | 'failed';
  progress_percent: number;
  vulns_found: number;
  message?: string;
  error?: string;
}> {
  return apiRequest(`/scan/${taskId}/status`);
}

export async function getSecurityScanResult(taskId: string): Promise<SecurityScanResult> {
  return apiRequest(`/scan/${taskId}/result`);
}

export async function listSecurityScans(): Promise<{
  success: boolean;
  scans: SecurityScanSummary[];
}> {
  return apiRequest('/scans');
}

export async function deleteSecurityScan(taskId: string): Promise<{ success: boolean }> {
  return apiRequest(`/scan/${taskId}`, { method: 'DELETE' });
}

export async function exportSecurityReport(taskId: string, format: 'json' | 'html'): Promise<{
  success: boolean;
  report_content?: string;
  report_url?: string;
  error?: string;
}> {
  return apiRequest(`/scan/${taskId}/export?format=${format}`, {
    method: 'GET',
  });
}

export async function testSecurityEndpoint(url: string): Promise<{
  success: boolean;
  status_code?: number;
  response_time_ms?: number;
  headers?: Record<string, string>;
  server?: string;
  powered_by?: string;
  error?: string;
}> {
  return apiRequest('/test-endpoint', {
    method: 'POST',
    body: JSON.stringify({ url }),
  });
}

export const SCAN_TYPE_OPTIONS = [
  { value: 'quick', label: '快速扫描', description: 'SQL注入、XSS基础检测，2-5分钟' },
  { value: 'full', label: '全面扫描', description: '全插件覆盖，包含SQL/XSS/XXE/Cmd注入等，10-30分钟' },
  { value: 'xss', label: 'XSS 专项', description: '仅检测跨站脚本漏洞，3-8分钟' },
  { value: 'sqli', label: 'SQL注入专项', description: '仅检测SQL注入漏洞，5-15分钟' },
  { value: 'cmd', label: '命令注入专项', description: '仅检测命令注入漏洞，2-5分钟' },
];

export const SEVERITY_COLORS: Record<string, string> = {
  critical: 'bg-red-500/20 text-red-400 border-red-500/30',
  high: 'bg-orange-500/20 text-orange-400 border-orange-500/30',
  medium: 'bg-yellow-500/20 text-yellow-400 border-yellow-500/30',
  low: 'bg-blue-500/20 text-blue-400 border-blue-500/30',
  info: 'bg-gray-500/20 text-gray-400 border-gray-500/30',
};
